File: yunva17/DFS&BFS/연구소.py
n, m = map(int, input().split())

graph = []
map = [[0]*m for _ in range(n)]
for _ in range(n):
    graph.append(list(map(int, input().split())))

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

result = 0

def add(x, y):
    for i in range(4):
        nx = x+dx[i]
        ny = y+dy[i]

        if nx>=0 and nx<n and ny>=0 and ny<m:
            if map[nx][ny] ==0:
                map[nx][ny]=2
                add(nx, ny)

def plus():
    score = 0
    for i in range(n):
        for j in range(m):
            if map[i][j] ==0:
                score+=1
    return score


# 매번 크기 계산
def dfs(count):
    global result
    if count == 3:
        for i in range(n):
            for j in range(m):
                map[i][j] = graph[i][j]
            for i in range(n):
                for j in range(m):
                    if map[i][j] ==2:
                        add(i,j)

            result = max(result, plus())
            return
    for i in range(n):
        for j in range(m):
            if graph[i][j] ==0:
                graph[i][j] =1
                count+=1
                dfs(count)
                graph[i][j]=0
                count-=1

dfs(0)
print(result)

File: jifrozen/DFS_BFS/연구소.java
package DFS_BFS;
import java.util.*;
import java.io.*;

public class 연구소 {
	static int n,m,result=0;
	static int[][] arr=new int[8][8];
	static int[][] temp=new int[8][8];
	static int[] dx={-1,0,1,0};
	static int[] dy={0,-1,0,1};
	public static void virus(int x,int y){
		for(int i=0;i<4;i++){
			int nx=x+dx[i];
			int ny=y+dy[i];
			if(nx>=0&&nx<n&&ny>=0&&ny<m){
				if(temp[nx][ny]==0){
					temp[nx][ny]=2;
					virus(nx,ny);
				}
			}
		}
	}

	public static int getScore(){
		int score=0;
		for(int i=0;i<n;i++){
			for(int j=0;j<m;j++) {
				if(temp[i][j]==0){
					score+=1;
				}
			}
		}
		return score;
	}

	public static void dfs(int count){
		if(count==3) {
			for (int i = 0; i < n; i++) {
				for (int j = 0; j < m; j++) {
					temp[i][j] = arr[i][j];
				}
			}
			for (int i = 0; i < n; i++) {
				for (int j = 0; j < m; j++) {
					if (temp[i][j] == 2) {
						virus(i, j);
					}
				}
			}
			result = Math.max(result, getScore());
			return;
		}
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				if(arr[i][j]==0){
					arr[i][j]=1;
					count+=1;
					dfs(count);
					arr[i][j]=0;
					count-=1;
				}

			}
		}
	}

	public static void main(String[] args) throws IOException{
		BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st=new StringTokenizer(br.readLine());

		n=Integer.parseInt(st.nextToken());
		m=Integer.parseInt(st.nextToken());

		for(int i=0;i<n;i++){
			st=new StringTokenizer(br.readLine());
			for(int j=0;j<m;j++){
				arr[i][j]=Integer.parseInt(st.nextToken());
			}
		}

		dfs(0);
		System.out.print(result);
	}
}
